src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+        MsiState {
+            msg_ctl,
+            ..Default::default()
+        }
+    }
+
+    /// Check whether the PCI MSI capability has been enabled.
+    pub fn enabled(&self) -> bool {
+        self.msg_ctl & MSI_CTL_ENABLE == MSI_CTL_ENABLE
+    }
+
+    /// Get number of vectors supported.
+    pub fn num_vectors(&self) -> u16 {
+        let field = (self.msg_ctl >> 1) & 0x7;
+
+        if field > 5 {
+            return 0;
+        }
+
+        1 << field
+    }
+
+    /// Get number of vectors enabled.
+    pub fn num_enabled_vectors(&self) -> u16 {
+        let field = (self.msg_ctl >> 4) & 0x7;
+
+        if field > 5 {
+            return 0;
+        }
+
+        1 << field
+    }
+
+    /// Handle update to the PCI MSI capability structure.
+    pub fn synchronize_state<I: InterruptManager>(
+        &mut self,
+        cap: &MsiCap,
+        intr_manager: &mut DeviceInterruptManager<I>,
+    ) -> std::io::Result<()> {
+        if self.msg_ctl != cap.msg_ctl {
+            self.update_msi_ctl(cap.msg_ctl, intr_manager)?;
+        }
+        if self.msg_addr_lo != cap.msg_addr_lo
+            || self.msg_data != cap.msg_data
+            || (cap.addr_64_bits() && self.msg_addr_hi != cap.msg_addr_hi)
+        {
+            self.msg_addr_lo = cap.msg_addr_lo;
+            self.msg_addr_hi = cap.msg_addr_hi;
+            self.msg_data = cap.msg_data;
+            self.update_msi_msg(intr_manager)?;
+        }
+        if cap.per_vector_mask() && self.mask_bits != cap.mask_bits {
+            self.update_msi_mask(cap.mask_bits, intr_manager)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn disable<I: InterruptManager>(
+        &mut self,
+        intr_manager: &mut DeviceInterruptManager<I>,
+    ) -> std::io::Result<()> {
+        if self.enabled() {
+            self.update_msi_ctl(self.msg_ctl & !MSI_CTL_ENABLE, intr_manager)?;
+        }
+
+        Ok(())
+    }
+
+    fn update_msi_ctl<I: InterruptManager>(
+        &mut self,
+        value: u16,
+        intr_manager: &mut DeviceInterruptManager<I>,
+    ) -> std::io::Result<()> {
+        match (self.enabled(), value & MSI_CTL_ENABLE != 0) {
+            (false, true) => {
+                if msi_num_enabled_vectors(value) > self.num_vectors() {
+                    debug!("guest OS enables too many MSI vectors");
+                } else {
+                    intr_manager.reset()?;
+                    intr_manager.set_working_mode(DeviceInterruptMode::PciMsiIrq)?;
+                    for idx in 0..self.num_enabled_vectors() as InterruptIndex {
+                        intr_manager.set_msi_high_address(idx, self.msg_addr_hi)?;
+                        intr_manager.set_msi_low_address(idx, self.msg_addr_lo)?;
+                        intr_manager.set_msi_data(idx, self.msg_data as u32 + idx)?;
+                        #[cfg(target_arch = "aarch64")]
+                        {
+                            intr_manager.set_msi_device_id(idx)?;
+                        }
```

Comment:
Why wrap this line in curly brackets? It is only one line.

Reply:
```
#[cfg(target_arch = "aarch64")]
                        {
                            intr_manager.set_msi_device_id(idx)?;
                        }
```
we would like to emphasize that this is an aarch64 method.